### ISSUE TITLE
New DataGenerator gets worse accuracy than old DataGenerator?

### ISSUE THREAD

**Issue by txizzle** [kratzert/finetune_alexnet_with_tensorflow#15]
Using the new ImageDataGenerator (June 15 update), I get lower accuracy than the old ImageDataGenerator (using cv2). 

I have changed the new ImageDataGenerator very slightly (tf.image.decode_png => tf.image.decode_jpeg, VGG_MEAN => IMAGENET_MEAN), but still get significant difference in accuracy. To benchmark the two datagenerators, I am loading in the BVLC AlexNet weights and validating on the ImageNet validation set without shuffle. 

With the old cv2 ImageDataGenerator, I get 55.72% Top-1 Accuracy and 79.08% Top-5 Accuracy. With the new TF ImageDataGenerator, I get 48.58% Top-1 Accuracy and 73.21% Top-5 Accuracy.

I suspect this has to do with how TF loads in images as opposed to CV2. Here is what the two different image processing steps look like:

Old CV2 ImageDataGenerator:
```
def next_batch(self, batch_size):
        """
        This function gets the next n ( = batch_size) images from the path list
        and labels and loads the images into them into memory 
        """
        # Get next batch of image (path) and labels
        paths = self.images[self.pointer:self.pointer + batch_size]
        labels = self.labels[self.pointer:self.pointer + batch_size]
        
        #update pointer
        self.pointer += batch_size
        
        # Read images
        images = np.ndarray([batch_size, self.scale_size[0], self.scale_size[1], 3])
        for i in range(len(paths)):
            img = cv2.imread(paths[i])
            
            #flip image at random if flag is selected
            if self.horizontal_flip and np.random.random() < 0.5:
                img = cv2.flip(img, 1)
            
            #rescale image
            img = cv2.resize(img, (self.scale_size[0], self.scale_size[1]))
            img = img.astype(np.float32)
            
            #subtract mean, which is np.array([104., 117., 124.])
            img -= self.mean
                                                                 
            images[i] = img

        # Expand labels to one hot encoding
        one_hot_labels = np.zeros((batch_size, self.n_classes))
        for i in range(len(labels)):
            one_hot_labels[i][labels[i]] = 1

        #return array of images and labels
        return images, one_hot_labels

```

New TF ImageDataGenerator:
```
def _parse_function_inference(self, filename, label):
        """Input parser for samples of the validation/test set."""
        # convert label number into one-hot-encoding
        one_hot = tf.one_hot(label, self.num_classes)

        # load and preprocess the image
        img_string = tf.read_file(filename)
        img_decoded = tf.image.decode_jpeg(img_string, channels=3)
        img_resized = tf.image.resize_images(img_decoded, [227, 227])
        
        IMAGENET_MEAN = tf.constant([104., 117., 124.], dtype=tf.float32)
        img_float = tf.to_float(img_resized)
        # RGB -> BGR
        img_bgr = img_float[:, :, ::-1]
        img_centered = tf.subtract(img_bgr, IMAGENET_MEAN)

        return img_centered, one_hot
```

I believe I am using the CV2 Datagenerator correctly as I can successfully finetune and call `train_generator.next_batch(batch_size)` within the training and validation loops with no issue. I built off of the finetune.py file you provided for the new TF Datagenerator (I use cpu device cpu:0 and then run the datagenerator init ops), and then get batches within the training/validation loops with `img_batch, label_batch = sess.run(next_batch)`. 

Any advice? Thanks a lot for providing and maintaining this proejct. Besides for this issue, it was super easy to work with and modify!

**Comment by kratzert** [comment]
Uh this is an interesting issue. I'll try to dig into this during the weekend. Out of curiosity, did you compared images loaded with the old datagenerator (with cv2) and the new (with TF)? I don't have time to look into this now, but I think I did compare the images loaded by both versions and they turned out the same. I suppose the rest of the script stays unchanged for your test? Can you provide the download link to the ImageNet validation set you used for testing, then I try to replicate this issue during the next days.

**Comment by txizzle** [comment]
The differences seem to be imperceptible to human eye but resulted in lower accuracy. Here's a [gist](https://gist.github.com/txizzle/6654a79086762fb75584fdbdbd0b2e04) with the code. It's an ipynb, but you should replace the filepath with your own path. Here's the image I demonstrated on:
![ilsvrc2012_val_00046916](https://user-images.githubusercontent.com/5369595/28990748-34306ac6-7935-11e7-8dbc-724629eab69b.jpeg)

Here's an SO question I posted: https://stackoverflow.com/questions/45516859/differences-between-cv2-image-processing-and-tf-image-processing

Cheers!

**Comment by kratzert** [comment]
I'll copy my answer from SO here:

As vijay m points out correctly, by changing the dct_method to "INTEGER_ACCURATE" you will get the same uint8 image using cv2 or tf. The problem indeed seems to be the resizing method. I also tried to force Tensorflow to use the same interpolation method than cv2 uses by default (bilinear) but the results are still different. This might be the case, because cv2 does the interpolation on integer values and TensorFlow converts to float before interpolating. But this is only a guess. If you plot the pixel-wise difference between the resized image by TF and cv2 you'll get the following historgram:

![figure_1](https://user-images.githubusercontent.com/13069767/28994890-ceeb6b8e-79d9-11e7-9caf-a5308580d54d.png)

As you can see, this looks pretty normal distributed. (Also I was surprised amount of pixel-wise difference). The problem of your accuracy drop could lie exactly here. In this paper Goodfellow et al. describe the effect of adversarial examples and classification systems. This problem here is something similar I think. If the original weights you use for your network were trained using some input pipeline which gives the results of the cv2 functions, the image from the TF input pipeline is something like an adversarial example. 

Example from the original publication:
![breakconv](https://user-images.githubusercontent.com/13069767/28994892-d52d071e-79d9-11e7-92c3-ed6b275903b2.png)

So in the end I think if you want to use the original network weights for the same data they trained the network on (ImageNet), you should stay with a similar/same input pipeline. If you use the weights to finetune the network on your own data, this should not be of a big concern, because you retrain the classification layer to work with the new input images (from the TF pipeline). And finetuning with both pipelines achieves roughly the same accuracy or me.

**Comment by txizzle** [comment]
Great, changing the `dct_method` did indeed fix the discrepancy between `cv2.imread() `and `tf.image.decode_jpeg()`. 

I see what you mean about the operations being different due to the difference between bilinear interpolation between integers (in `cv2`) and floats (in `TF`). However, we can test this by just converting the `cv2` image to a float before `cv2.resize()`:

```
# Read .JPEG image
cv2_decoded = cv2.imread(filepath)

# Convert to float and rescale image
cv2_float = cv2_decoded.astype(np.float32)
cv2_resized = cv2.resize(cv2_float, (227, 227))

# Subtract Imagenet mean
cv2_centered = cv2_resized - mean
```

This does effect the `cv2.resize()` result, but still does not match the `tf.image.resize_images()` result (when both methods are given the same input: (333,500,3) float32 with the same values).

The new results are:

```
tf_img_resized (convert to float then resize): 
dtype: float32
shape: (227, 227, 3)
[[  26.           25.82378769   35.13656616 ...,   80.68237305   59.6477356
    49.82373047]
 [  39.40088272   34.68881607   51.59126282 ...,   84.75141907
    75.70881653   23.05552483]
 [  51.27312851   29.41714287   39.21158981 ...,   67.73635101
    81.68902588   36.37636185]
 ..., 
 [  70.19384766   76.590271     46.52557755 ...,   70.67378998
    89.23838806   99.26979828]
 [  71.5375061    59.35364914   49.22055054 ...,   75.69091034
    99.11106873  105.69988251]
 [  65.           60.30866623   57.18851852 ...,   77.46831512
    99.54577637  106.68299866]]

cv2_resized:
dtype: float32
(227, 227, 3)
[[  36.4573555    29.99937057   33.85260773 ...,  101.9519043    59.46978378
    43.46925354]
 [  34.9861145    28.48155403   50.77778625 ...,   85.33702087
    60.68599701   25.6464119 ]
 [  28.01940727   39.1032753    49.93264389 ...,   59.17743683
    61.94624329   52.26992035]
 ..., 
 [  74.94075012   67.4425354    33.72472382 ...,   74.62219238
    97.91424561  100.79891968]
 [  67.43686676   58.67483521   43.38725281 ...,   86.18499756
   102.63587952  104.28365326]
 [  65.64103699   65.21386719   48.20243073 ...,   86.14474487
   103.37129974  104.63876343]]
```

Before casting the input of `cv2.resized()` to a float, the result was:

```
cv2_resized:
dtype: uint8
(227, 227, 3)
[[ 36  30  34 ..., 102  59  43]
 [ 35  28  51 ...,  85  61  26]
 [ 28  39  50 ...,  59  62  52]
 ..., 
 [ 75  67  34 ...,  74  98 101]
 [ 67  59  43 ...,  86 102 104]
 [ 66  65  48 ...,  86 103 105]]
```

So basically, I think there still might be something else beyond the automatic float conversion that is affecting the differences between `TF` and `cv2` resize operations. 

On the other hand, I do agree with you, that if we just accept that the `resize()` operations are going to be different between `TF` and `cv2`, a solution is just to finetune using the new pipeline to overcome the adversarial perturbations. Unfortunately, I am still working with the ImageNet dataset, so I am not sure if finetuning using the new pipeline will work well. Seems that we would be "unlearning" a lot of the knowledge learned during the original training phase using the cv2-like pipeline.

**Comment by kratzert** [comment]
This is indeed interesting, since they both claim to use the same method for interpolation. You could out of curiosity try how the different methods behave.
[TF methods](https://www.tensorflow.org/api_docs/python/tf/image/resize_images)
[OpenCV methods](http://docs.opencv.org/3.0-beta/modules/imgproc/doc/geometric_transformations.html#resize)

If it's only between linear interpolation, there might be an error in the TF or OpenCV method and we should try to use other libraries (maybe skimage, PIL) and see what there output is.

**Comment by txizzle** [comment]
Great news! Although I didn't figure out the issue with `tf.image.resize_images`, I did fix this particular issue of getting lower validation accuracy. Turns out the latest version of the `AlexNet` architecture in this project flipped the LRN and Max-pooling layers from the old version. 

So after swapping those layers, I was able to achieve the same validation accuracy as the old `cv2` version. This intuitively makes sense, because the learned model should not be so sensitive to random perturbations; I would have expected adversarial samples to be crafted fairly carefully, so it would be surprising that such a small change (two resize operations) would cause such a large perturbation. 

Anyways, was there any reason you switched the LRN and max-pooling layers?

**Comment by kratzert** [comment]
Hmm I can't see any commit where I changed the AlexNet architecture. But strangly I seemed to have been confused between AlexNet and the CaffeNet ref. version of AlexNet ([Link](https://github.com/BVLC/caffe/tree/master/models/bvlc_reference_caffenet)). I guess because nobody ever played with the original ImageNet validation data, this never came to light. Thanks for this hint, I'll swap the layer in the code.

So I didn't do any layer switch on purpose. Must have been confused between the Caffe version and the original network at some point. But since this is about AlexNet, I should change it.

I'm not sure if your assumption about the network sensitivity is right, though. I have never made experiments in this direction but this is exactly what the paper from above claimed.
